File: script/simple_stitching.py
#!/usr/bin/env python3
import rospy
import cv2
import numpy as np
import math
import rosparam
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

class simple_stitching:
  def __init__(self):
    try:
        sub_image1_topic_name = rosparam.get_param("theta_simple_stitching/sub_image1_topic_name")
        sub_image2_topic_name = rosparam.get_param("theta_simple_stitching/sub_image2_topic_name")
    except:
        rospy.logwarn("subscribe /image_raw because rosparam is not set")
        sub_image1_topic_name = "/image1_raw"
        sub_image2_topic_name = "/image2_raw"
    try:
        self.reverse = rosparam.get_param("theta_simple_stitching/reverse")
    except:
        rospy.logwarn("image is not reversed")
        self.reverse = False

    self.image_pub = rospy.Publisher("/image/mercator", Image, queue_size=1)
    self.bridge = CvBridge()
    self.image1_sub = rospy.Subscriber(sub_image1_topic_name, Image, self.image1_callback)
    self.image2_sub = rospy.Subscriber(sub_image2_topic_name, Image, self.image2_callback)
    self.image_s1 = np.array
    self.image_s2 = np.array

    vertex = 640
    src_cx = 319
    src_cy = 319
    src_r = 319 #283
    src_cx2 = 1280 - src_cx

    self.map_x = np.zeros((vertex,vertex*2))
    self.map_y = np.zeros((vertex,vertex*2))
    for y in range(vertex):
        for x in range(vertex*2):
            phi1 = math.pi * x / vertex
            theta1 = math.pi * y / vertex

            X = math.sin(theta1) * math.cos(phi1)
            Y = math.sin(theta1) * math.sin(phi1)
            Z = math.cos(theta1)

            phi2 = math.acos(-X)
            theta2 = np.sign(Y)*math.acos(-Z/math.sqrt(Y*Y + Z*Z))

            #0 equidistant projection
            #1 stereographic projection
            #2 stereographic inverse projection
            #3 Orthogonal projection
            #4 Orthogonal inverse projection
            method = 0

            if(phi2 < math.pi / 2):
                if method == 0:
                    r_ = phi2 / math.pi * 2
                elif method == 1:
                    r_ =  math.tan((phi2) / 2)
                elif method == 2:
                    r_ =  1 - math.tan((math.pi / 2 - phi2) / 2)
                elif method == 3:
                    r_ = math.sin(phi2)
                elif method == 4:
                    r_ = 1 - math.sin(math.pi / 2 - phi2)
                self.map_x[y,x] = src_r * r_ * math.cos(theta2) + src_cx
                self.map_y[y,x] = src_r * r_ * math.sin(theta2) + src_cy
            else:
                if method == 0:
                    r_ = (math.pi - phi2) / math.pi * 2
                elif method == 1:
                    r_ =  math.tan((math.pi - phi2) / 2)
                elif method == 2:
                    r_ =  1 - math.tan((-math.pi/2 + phi2) / 2)
                elif method == 3:
                    r_ = math.sin(math.pi - phi2)
                elif method == 4:
                    r_ = 1 - math.sin(- math.pi / 2 + phi2)
                self.map_x[y,x] = src_r * r_ * math.cos(math.pi - theta2) + src_cx2
                self.map_y[y,x] = src_r * r_ * math.sin(math.pi - theta2) + src_cy

    self.map_x = self.map_x.astype('float32')
    self.map_y = self.map_y.astype('float32')
    print("finish init for theta simple stitching")


  def image1_callback(self,data):
    try:
      image1 = self.bridge.imgmsg_to_cv2(data, "bgr8")
    except CvBridgeError as e:
      print(e)

    self.image_s1 = cv2.resize(image1, (640,640))

  def image2_callback(self,data):
    try:
      image2 = self.bridge.imgmsg_to_cv2(data, "bgr8")
    except CvBridgeError as e:
      print(e)

    self.image_s2 = cv2.resize(image2, (640,640))
    
  def stitching(self):
    image_cat = cv2.hconcat([self.image_s1, self.image_s2])
    image_remap = cv2.remap( image_cat, self.map_x, self.map_y, cv2.INTER_LINEAR, cv2.BORDER_CONSTANT);
    if(self.reverse):
        tmp = image_remap.copy()
        image_remap[:, :1280/2,:] = tmp[:, 1280/2:,:]
        image_remap[:, 1280/2:,:] = tmp[:, :1280/2,:]

    msg_for_send = self.bridge.cv2_to_imgmsg(image_remap, "bgr8")
    # msg_for_send.header = Image.header

    try:
      self.image_pub.publish(msg_for_send)
    except CvBridgeError as e:
      print(e)

if __name__ == '__main__':
    rospy.init_node('theta_simple_stitching')
    ss = simple_stitching()
    rospy.wait_for_message("/camera/rgb/image_raw_front", Image)
    rate = rospy.Rate(10)
    while not rospy.is_shutdown():
        ss.stitching()
        rate.sleep()
        

File: src/simple_stitching.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
 
class theta_simple_stitching{
public:
  theta_simple_stitching();
  void imageCb(const sensor_msgs::ImageConstPtr& msg_ptr);
private:
  ros::NodeHandle nh_;
  ros::NodeHandle pnh_;
  image_transport::ImageTransport it_;
  image_transport::Subscriber image_sub_;
  image_transport::Publisher image_pub_;
  cv::Mat map_x_;
  cv::Mat map_y_;
  int method_;
  double vertex_;
  double src_cx_;
  double src_cy_;
  double src_r_;
  double src_cx2_;
};

theta_simple_stitching::theta_simple_stitching() :
    nh_(),
    it_(nh_),
    pnh_("~"),
    method_(0),
    vertex_(540),
    src_cx_(319),
    src_cy_(319),
    src_r_(283),
    src_cx2_(1280 - src_cx_)
  {
    image_sub_ = it_.subscribe("/image_raw", 1, &theta_simple_stitching::imageCb, this);
    image_pub_ = it_.advertise("/image/mercator", 1);

    map_x_ = cv::Mat::zeros(vertex_, vertex_*2, CV_32FC1);
    map_y_ = cv::Mat::zeros(vertex_, vertex_*2, CV_32FC1);
    for(int y; y < vertex_; y++){
      for(int x; x < vertex_; x++){
        double phi1 = M_PI * x / vertex_;
        double theta1 = M_PI * y / vertex_;
        double X = std::sin(theta1) * std::cos(phi1);
        double Y = std::sin(theta1) * std::sin(phi1);
        double Z = std::cos(theta1);
        double phi2 = std::acos(-X);
        double sign = 0;
        if(Y > 0) sign = 1;
        else if(Y < 0) sign = -1;
        double theta2 = sign * std::acos(-Z/std::sqrt(Y*Y + Z*Z));
        double r_ = 0;
        if(phi2 < M_PI_2){
          switch(method_){
            case 0:
              r_ = phi2 / M_PI * 2;
              break;
            case 1:
              r_ = std::tan(phi2 / 2);
              break;
            case 2:
              r_ = 1 - std::tan((M_PI_2 - phi2) / 2);
              break;
            case 3:
              r_ = std::sin(phi2);
              break;
            case 4:
              r_ = 1 - std::sin(M_PI_2 - phi2);
              break;
          }
          cv::Vec3b *src;
          src = map_x_.ptr<cv::Vec3b>(y);
          src[x] = src_r_ * r_ * std::cos(M_PI - theta2) + src_cx2_;
          src = map_y_.ptr<cv::Vec3b>(y);
          src[x] = src_r_ * r_ * std::sin(M_PI - theta2) + src_cx2_;
        }
        else{
          switch(method_){
            case 0:
              r_ = (M_PI - phi2) / M_PI * 2;
              break;
            case 1:
              r_ = std::tan(M_PI - phi2 / 2);
              break;
            case 2:
              r_ = 1 - std::tan((M_PI_2 + phi2) / 2);
              break;
            case 3:
              r_ = std::sin(M_PI + phi2);
              break;
            case 4:
              r_ = 1 - std::sin(-M_PI_2 + phi2);
              break;
          }
          cv::Vec3b *src;
          src = map_x_.ptr<cv::Vec3b>(y);
          src[x] = src_r_ * r_ * std::cos(M_PI - theta2) + src_cx2_;
          src = map_y_.ptr<cv::Vec3b>(y);
          src[x] = src_r_ * r_ * std::sin(M_PI - theta2) + src_cx2_;
        }
        ROS_INFO("finish init for theta siimple stitching");
      }
    }
  }

  void theta_simple_stitching::imageCb(const sensor_msgs::ImageConstPtr& msg_ptr){
    cv::Mat cv_image;
    try{
      cv_image = cv_bridge::toCvCopy(msg_ptr, sensor_msgs::image_encodings::BGR8)->image;
    }
    catch (cv_bridge::Exception& e){
      ROS_ERROR("cv_bridge exception: %s", e.what());
      return;
    }
    cv::Mat resized_image;
    cv::resize(cv_image, resized_image, cv::Size(1280, 720), cv::INTER_LINEAR);

    cv::Mat result_image;
    cv::remap(resized_image, result_image, map_x_, map_y_, cv::INTER_LINEAR, cv::BORDER_CONSTANT);

    sensor_msgs::ImagePtr pub_msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", result_image).toImageMsg();
    image_pub_.publish(pub_msg);
}

 
int main(int argc, char** argv){
  ros::init(argc, argv, "theta_simple_stitching");
  theta_simple_stitching ss;
  ros::spin();
  return 0;
}
